src/state.rs

```diff
@@ -365,4 +367,519 @@ mod tests {
+
+                            // if our Acks are not done, we will not understand this vote
+                            if !nodes[to].acks_done() {
+                                resp.push(MockVoteResponse::RequestAntiEntropy);
+                            } else if let Some(val) = res {
+                                if nodes[to].all_acks_done() {
+                                    resp.push(MockVoteResponse::DkgComplete)
+                                } else {
+                                    // if false, we need more votes
+                                    if !val {
+                                        resp.push(MockVoteResponse::WaitingForMoreVotes)
+                                    }
+                                }
+                            };
+                            (to, resp)
+                        })
+                        .collect();
+
+                    commands.push(SendVote::AllAcks(current_node, to_nodes_resp));
+                }
+                _ => {}
+            }
+
+            active_nodes = MockNode::active_nodes(&nodes);
+        }
+        commands
+    }
+
+    // Test helpers
+    fn generate_nodes<R: RngCore>(
+        num_nodes: usize,
+        threshold: usize,
+        mut rng: &mut R,
+    ) -> Result<Vec<DkgState>> {
+        let secret_keys: Vec<SecretKey> = (0..num_nodes).map(|_| bls::rand::random()).collect();
+        let pub_keys: BTreeMap<_, _> = secret_keys
+            .iter()
+            .enumerate()
+            .map(|(id, sk)| (id as u8, sk.public_key()))
+            .collect();
+        secret_keys
+            .iter()
+            .enumerate()
+            .map(|(id, sk)| {
+                DkgState::new(id as u8, sk.clone(), pub_keys.clone(), threshold, &mut rng)
+            })
+            .collect()
+    }
+
+    #[derive(Debug)]
+    enum SendVote {
+        // from_node, list of (to_node, vec of response when handled)
+        Parts(usize, Vec<(usize, Vec<MockVoteResponse>)>),
+        Acks(usize, Vec<(usize, Vec<MockVoteResponse>)>),
+        AllAcks(usize, Vec<(usize, Vec<MockVoteResponse>)>),
+    }
+
+    #[derive(Debug)]
+    enum MockVoteResponse {
+        WaitingForMoreVotes,
+        BroadcastVote(MockDkgVote),
+        RequestAntiEntropy,
+        DkgComplete,
+    }
+
+    impl MockVoteResponse {
+        pub fn match_resp(
+            &self,
+            actual_resp: VoteResponse,
+            update_acks: &mut BTreeMap<usize, DkgSignedVote>,
+            update_all_acks: &mut BTreeMap<usize, DkgSignedVote>,
+            update_sk: &mut BTreeMap<usize, SecretKeyShare>,
+            update_pk: &mut BTreeSet<PublicKeySet>,
+            id: usize,
+        ) -> bool {
+            if (matches!(self, Self::WaitingForMoreVotes)
+                && matches!(actual_resp, VoteResponse::WaitingForMoreVotes))
+                || (matches!(self, Self::RequestAntiEntropy)
```

Comment:
I wonder if we could impl a PartialEq trait between the mock type and the actual type? This would make comparisons simpler.

Reply:
Yeah!! You're right! Pushed a much simpler code!